pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/purge/PurgeTaskGenerator.java

```diff
@@ -0,0 +1,168 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.plugin.minion.tasks.purge;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import org.apache.pinot.common.data.Segment;
+import org.apache.pinot.common.metadata.segment.SegmentZKMetadata;
+import org.apache.pinot.controller.api.exception.UnknownTaskTypeException;
+import org.apache.pinot.controller.helix.core.minion.generator.BaseTaskGenerator;
+import org.apache.pinot.controller.helix.core.minion.generator.TaskGeneratorUtils;
+import org.apache.pinot.core.common.MinionConstants;
+import org.apache.pinot.core.minion.PinotTaskConfig;
+import org.apache.pinot.spi.annotations.minion.TaskGenerator;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableTaskConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@TaskGenerator
+public class PurgeTaskGenerator extends BaseTaskGenerator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PurgeTaskGenerator.class);
+  private static final String DEFAULT_DELTA_PERIOD = "1d";
+
+  @Override
+  public String getTaskType() {
+    return MinionConstants.PurgeTask.TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
+    LOGGER.info("Start generating PurgeTask");
+    String taskType = MinionConstants.PurgeTask.TASK_TYPE;
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+
+    for (TableConfig tableConfig : tableConfigs) {
+
+      String tableName = tableConfig.getTableName();
+      if (tableConfig.getTableType() == TableType.REALTIME) {
+        LOGGER.info("Task type : {}, cannot be run on table of type  {}", taskType, TableType.REALTIME);
+        continue;
+      }
+
+      Map<String, String> taskConfigs;
+      try {
+        TableTaskConfig tableTaskConfig = tableConfig.getTaskConfig();
+        Preconditions.checkNotNull(tableTaskConfig);
+        taskConfigs = tableTaskConfig.getConfigsForTaskType(MinionConstants.PurgeTask.TASK_TYPE);
+        Preconditions.checkNotNull(taskConfigs, "Task config shouldn't be null for Table: {}", tableName);
+      } catch (Exception e) {
+        continue;
+      }
+      String deltaTimePeriod =
```

Comment:
You don't need to introduce a new concept of `DELTA_TIME_PERIOD_KEY` to schedule the purge task. Instead the purge task should be set up on segment level. And since the number of segments is different for different tables, this new config will schedule unbalanced purge tasks per day. So I suggest removing this new table level `DELTA_TIME_PERIOD_KEY`.
Some variable called `lastPurgeTime` can be calculated for each of the segments. By default the value should be the segment creation time. If there is a field called "PurgeTask.time" in the custom map in the ZK metadata, it should override the default value.
You can leave a TODO in this generator, and I'll fill the logic once this PR is merged.

Reply:
I've implement the threesold asked here : https://github.com/apache/pinot/pull/8589#discussion_r864089515 .

Reply:
@jackjlli This delta is essentially the time threshold to skip purging if the segment is just purged within the threshold (e.g. 2 weeks for GDPR requirement)